serialization_v2/serialization.py

```diff
@@ -0,0 +1,1280 @@
+                    if assign.ref_count > 1:
+                        # there are multiple references, so we need to keep this assignment
+                        return PyEvalCode(assign.py_name)
+                    if not assign.py_value_repr:
+                        return PyEvalCode(assign.py_name)  # we cannot inline this, so just use the assignment
+                    # We can inline this.
+                    # Thus remove the reference to this assignment.
+                    assign.ref_count -= 1
+                    assert assign.ref_count == 0
+                    # Can delete this assignment.
+                    del self.assignments_dict_by_value_ref[value_ref]
+                    del self.assignments_dict_by_name[assign.py_name]
+                    del self.assignments_dict_by_idx[assign.idx]
+                    return assign.py_value_repr
+            else:
+                assign.ref_count += 1
+                self._cur_added_refs.append(assign)
+                return PyEvalCode(assign.py_name)
+        if not self._inlining_stage:
+            value_dict = self.assignments_dict_by_value_by_type.get(type(value))
+            if value_dict is not None and value in value_dict:
+                assign = value_dict.get(value)
+                if assign is not None:
+                    assign.ref_count += 1
+                    self._cur_added_refs.append(assign)
+                    return PyEvalCode(assign.py_name)
+
+        # Any of the following could potentially cause further recursive calls,
+        # thus check the recursive flag at this point.
+        if recursive:
+            assert not self._inlining_stage  # should not get here when inlining
+            raise _SerializationDependsOnNotYetSerializedOtherVarException(
+                _AssignQueueItem(value=value, suggested_var_name=prefix),
+            )
+        assert name
+
+        if isinstance(value, dict):
+            return self._serialize_dict(value, prefix)
+        if isinstance(value, list):
+            return self._serialize_list(value, prefix)
+        if isinstance(value, tuple):
+            return self._serialize_tuple(value, prefix)
+        if isinstance(value, set):
+            return self._serialize_set(value, prefix)
+        if _isinstance_returnn_dim(value):
+            return self._serialize_dim(value, prefix)
+        if _isinstance_returnn_cached_file(value):
+            return self._serialize_cached_file(value, prefix)
+        if isinstance(value, functools.partial):
+            return self._serialize_functools_partial(value, name)
+        if isinstance(value, Call):
+            return self._serialize_call(value, name)
+        if isinstance(value, CallImport):
+            return self._serialize_call_import(value, name)
+        if isinstance(value, PartialImport):
+            return self._serialize_partial_import(value, name)
+        if isinstance(value, Import):
+            return self._serialize_import(value, name)
```

Comment:
Despite the discussion whether you should or should not use those here for `ReturnnConfigWithNewSerialization`, I wonder, why do you need to have this code here at all? Will you even ever get here? Because you have called `instanciate_delayed_copy` already on the config, which should have resolved all of that, or not?

Reply:
See https://github.com/rwth-i6/i6_core/pull/601#issuecomment-3540972389

Reply:
To answer myself: This is because the `instanciate_delayed_copy` was also modified to be inconsistent to `instanciate_delayed` in that it would *not* handle `SerializerObject`.